Populate config options to node-local networks

stffabi opened the pull request:
Populate driver and IPAM config options for node-local networks to the network runtime configs, this makes sure the networks get created with all options on the swarm nodes.

Fixes moby/moby#37089
Fixes moby/moby#36964

**- What I did**

Fixed the swarm network creation to use the correct IPAM and driverConfig settings.

**- How I did it**

For node-local networks, e.g. the bridge network, the swarm manager doesn't allocate anything because the driver doesn't support that. Therefore the runtime options for the driver and the IPAM configs are never populated. As those runtime options are used for creating the network later on the node, the setting previously provided to "docker network create" have been lost and won't be applied. This PR additionally populates the driver options and IPAM settings to the runtime settings for node-local networks, which will then be propagated to the nodes. As long as there's no way to allocate something for node-local networks on the swarm manager, it should at least propagate the settings over to the nodes. I hope this is the correct solution, what do you think @abhi ? Thanks again for your input on this issue. If you @abhi are okey with this solution just let me know and I'm going to also add some unit tests for this.

**- How to verify it**

See reproduction instructions on moby/moby#37089

**- Description for the changelog**

Populate driver and IPAM config options for node-local networks to the network runtime configs, this makes sure the networks get created with all options on the swarm nodes.

stffabi commented:
Ping @abhi, would it be possible for you to take a look into this?

stffabi commented:
@thaJeztah would it be possible for you to ping someone internally at docker to look into this PR? That would be really awesome 😃, because our company is currently a little bit blocked due to this issue.

Thanks to all of the docker team, for the hard work.

abhi commented on manager/allocator/cnmallocator/networkallocator.go:
```diff
@@ -139,16 +139,39 @@ func (na *cnmNetworkAllocator) Allocate(n *api.Network) error {
+		if n.Spec.IPAM != nil {
+			if n.Spec.IPAM.Driver != nil {
+				n.IPAM.Driver.Name = n.Spec.IPAM.Driver.Name
+				n.IPAM.Driver.Options = n.Spec.IPAM.Driver.Options
```
Please follow ```allocateDriverState``` to set the options correctly.

stffabi replied:
Ah I see, you mean the combination of the probably received operational state from the store with the network spec.

abhi commented on manager/allocator/cnmallocator/networkallocator.go:
```diff
@@ -139,16 +139,39 @@ func (na *cnmNetworkAllocator) Allocate(n *api.Network) error {
+			ipamSpec := make([]*api.IPAMConfig, 0, len(n.Spec.IPAM.Configs))
+			for _, ic := range n.Spec.IPAM.Configs {
+				ipamSpec = append(ipamSpec, &api.IPAMConfig{
+					Subnet:  ic.Subnet,
```
Isnt subnet config local to node ? It doesnt have to be set at global level here for node local networks. I think waht you need is only options

stffabi replied:
Thanks @abhi  for looking into it.

Yes, subnet config should be local to node, which means every node should have the same subnet. E.g. creating a network like this `docker network create --driver=bridge --subnet=192.168.0.0/24 --scope=swarm test` should result in bridge networks with 192.168.0.0/24 on every node. From the point of view of the network allocator it's a global config, which needs to be passed to every node, like for other non-node local networks also.

Subnet and gateway config is not stored in the options AFAIK, therefore leaving away those configs would result in a bridge network on every node without the specified subnet. It won't also be displayed by docker network inspect run on a master node.

stffabi commented:
@abhi  I've updated the PR and split out some code from `allocatePools` and `allocateDriverState`, so now both node-local and global network use the same code to determine the initial driver and ipam state.

stffabi commented:
FYI: @thaJeztah , @abhi 
In the meantime our company managed to work around this issue by implementing our docker network driver to create non node-local networks. Therefore this isn't critical for us anymore and as communication flow was/is very stagnant I assume docker is not very interested in this PR and it has not a very high position on the todo list of docker. 

So I'll go ahead and close the PR...
